### DISCUSSION TITLE
Metadata usage and handling

### DISCUSSION THREAD

**Discussion by pjotrek-b** [minio/minio#19801] [Q&A] [answered]
Hello everyone :smile: 

I'm very new to this. I'm an experienced system-and-network admin, and I've setup quite a lot of exciting constellations of small-to-large-scale network storages.

From reading articles and documentation and watching tutorials, I understood that minio is designed/used to do the following:

  * store objects, *not* files in folders
  * store,access and query (?) metadata
  * store meta with data (as object)

Now when I run my hello-world Python API code, I can create, access - and even set metadata and tags of an Object.

However, I'd be very interested at more real-world examples, I've missed in existing online material.

Like:
**Using a key=value exiftool-extracted MP3 metadata in plain text, storing each line as custom Object-Metadata**

I wrote code to do that, but then get server 400 errors by Minio unless I completely strip and truncate the real-world random audiofile mp3 tags. This seems surprising to me, as I've read it should support UTF-8?

I'm very happy to exchange experiences and ideas about object-meta handling!
And grateful for any suggestions or insights.

**Comment by pjotrek-b** [comment]
I have on purpose decided not (yet) to publish my code and its output. I can post it here upon interest :)

**Comment by harshavardhana** [comment, marked as answer]
> as I've read it should support UTF-8?

UTF-8 is not supported for tags

**Reply by pjotrek-b** [reply]
Oh.
I was referring to : https://docs.aws.amazon.com/tag-editor/latest/userguide/tagging.html
Which says:

* The tag key must be a minimum of 1 and a maximum of 128 Unicode characters in UTF-8.
* The tag value must be a minimum of 0 and a maximum of 256 Unicode characters in UTF-8.

oh, and btw: mentions *50 user created tags* as limit now (I'm sure it was 10).

**Reply by harshavardhana** [reply]
That is meant for AWS EC2 instances, AWS S3 however has much lower limits

**Reply by harshavardhana** [reply]
https://github.com/minio/minio-go/blob/master/pkg/tags/tags.go#L72

**Reply by pjotrek-b** [reply]
> That is meant for AWS EC2 instances, AWS S3 however has much lower limits

Oh, I see.
Thank you very much for the link to where I can increase the AWS limits for tag number and size.
I'll try to compile minio with changed values on a VM.

Where could I find the documentation about "allowed characters" in tags *and metadata* on minio?

**Reply by pjotrek-b** [reply]
> https://github.com/minio/minio-go/blob/master/pkg/tags/tags.go#L72

oh, I realized that's the s3 go-client - not the minio server side.
Would it suffice to change only the client side, or does the minio server also enforce those limitations?

Anyhow:
I feel I've misunderstood how Object Storage is to be used, when it comes to meta+data handling - otherwise I wouldn't be the only one being puzzled by how restrictive the "meta" storage-and-handling seems to be.

I'm looking for reference implementations/code examples of serious *metadata* handling with Minio, but cannot really find any.
